docs/adr/007_cups_firmware_upgrade.md

```diff
@@ -0,0 +1,129 @@
+- Defining the changes required in the storage solution used for the CUPS
+  protocol implementation
+- Defining the changes needed in the Azure Function for supporting firmware
+  upgrades
+- Defining the changes needed in LoRaWan Network Server (LNS) for handling
+  firmware upgrades
+- Defining the changes needed in LoRa Device Provisioning CLI for allowing
+  firmware upgrades
+
+## Out-of-scope
+
+## Decisions
+
+### Context
+
+The CUPS request described in [The CUPS protocol - documentation][cupsproto]
+contains the `package` field which indicates the current firmware version of the
+Basics Station. The value will need to be updated whenever a firmware upgrade of
+the Basics Station is performed by the user.
+
+### Firmware upgrade flow
+
+1. The `package` parameter, along with URL pointing to the location of the
+   firmware upgrade binary in a storage account, will be stored in the desired
+   properties of the device twin of the concentrator running the Basics Station.
+   Those values reflect the up-to-date firmware version that needs to be used
+   when communicating with the CUPS server. The firmware upgrade is initiated
+   through updating these parameters in the concentrator device twin.
+
+1. The Basics Station sends the CUPS request containing the currently used
+   values.
+
+1. The LNS determines if there are discrepanices between the values stored in
+   the concentrator twin and the ones provided by the Station. If there are
+   differences, the LNS will trigger the firmware upgrade by sending the correct
+   values to the Basics Station. The Station will then execute the actual
+   firmware upgrade.
+
+### IoT Hub related changes
+
+Only change is related to the concentrator device twin.
+
+In addition to the values already stored in the twin, the following need to be
+added:
+
+```json
+"cups": {
+  // ...
+  "package": "1.0.1",
+  "fwUrl": "https://...",
+  "fwKeyChecksum": 123456,
+  "fwSignature": ""
+}
+```
+
+- **'package'**: current package version of the Station
+- **'fwUrl'**: URL pointing to the storage location of the file required to run
+  the upgrade
+- **'fwKeyChecksum'**: checksum of the key used to sign the digest of the
+  firmware upgrade file
+- **'fwSignature'**: signature of the uploaded firmware upgrade file
+
+### Storage related changes
+
+We will use the same storage solution as the one selected for the general CUPS
+protocol support (storage account). A new container will be used to store the
+firmware upgrade files.
+
+### Azure Function related changes
+
+A new endpoint will be added in the Facade Azure Function which will be used to
+fetch firmware upgrade files from the storage account.
+
+### Changes in the LoRaWan Network Server
+
+The implementation of `CupsProtocolMessageProcessor` should be extended for
+checking the `package` field from the concentrator device twin. In case there
+the value is different from the one received from the Basics Station in the CUPS
```

Comment:
If the value of 'package' field is different, we have to first retrieve the twin and check whether any of the key in the keys array from the request is actually equal to the fwKeyChecksum field that we have in the twin.

If there is no matching checksum, it means that the concentrator is missing the key required for calculating the digest and verifying the update file. When this happens we should throw an appropriate error